hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java

```diff
@@ -1136,9 +1137,25 @@ public R getSingleResult() {
 	// Named query externalization
 
 	@Override
-	public NamedHqlQueryMemento toMemento(String name) {
+	public NamedQueryMemento toMemento(String name) {
 		if ( CRITERIA_HQL_STRING.equals( getQueryString() ) ) {
-			throw new UnsupportedOperationException( "Criteria-based Query cannot be saved as a named query" );
+			return new NamedSqmQueryMementoImpl(
+					name,
+					getSqmStatement(),
```

Comment:
Dunno how "safe" we want this to be, but maybe we should copy the statement here?

```suggestion
					getSession().isJpaCriteriaCopyComplianceEnabled() ? getSqmStatement() : getSqmStatement().copy( SqmCopyContext.simpleContext(),
```